### DISCUSSION TITLE
Rust projects with multiple targets

### DISCUSSION THREAD

**Discussion by tgecho** [zed-industries/zed#11750] [Feature Requests]
I work on some code bases that target wasm and native targets. ~95% of the code is shared, but there are a few key modules/crates that are split using `#[cfg(target_arch = "wasm32")]` style attributes.

In vscode, I can have separate `native.code-workspace` and `wasm.code-workspace` files with something like 
```
{
  "settings": {
    "rust-analyzer.cargo.target": "wasm32-unknown-unknown"
  }
}
```

This lets me have two different vscode workspaces open at the same time and jump between them if I need to focus on code specific to a given target.

Is there any way to accomplish something similar in zed?

I found this issue, which would give me a direct equivalent, but I'm curious if there's some other way of doing things that might work right now: [Support opening .code-workspace files to open vs code type workspace directly](https://github.com/zed-industries/zed/issues/9459)

This also seems related: [
Configure rust toolchain on a per-folder/project level
](https://github.com/zed-industries/zed/issues/4883)

**Comment by osiewicz** [comment]
Is there anything distinctive in the structure of your code base that would allow us to pick up on different toolchain being applied to the WASM part? Do you e.g. have a `rust-toolchain.toml` file?

We could probably make it work with toolchains, which landed a few months back.

**Reply by tgecho** [reply]
I don't think a config file in the code base can really cover my use case (though I'm extremely open to suggestions for alternate ways to structure things!).

Some of my crates are WASM specific, some are native specific, and some are used in both contexts. I try to keep the shared crates as target neutral as possible, but there are places where I just have to implement alternate code paths with cfg attributes:

```rust
#[cfg(not(target_arch = "wasm32"))]
struct Foo {
	something: ThatDoesNotWorkWithWasm
}

#[cfg(target_arch = "wasm32")]
struct Foo {
	something: WasmSpecific
}
```

Since these are mutually exclusive, as far as rust-analyzer is concerned basically only one of them exists at a time. I get no checking/actions/etc on the target that isn't active. Using a vscode workspace per target lets me work in "WASM mode" or "native mode" without modifying a config file.

**Reply by osiewicz** [reply]
That's fine. I've suggested using `rust-toolchain.toml` since it is a "standard" file that lets you specify the toolchains that are necessary to work with a given project (see [our file](https://github.com/zed-industries/zed/blob/main/rust-toolchain.toml) for an example). We can always resort to listing all available toolchains, that's easy peasy - but there may also be lots of them, which is why I think we should list ones in `rust-toolchain.toml` as ones at the top, as they're more likely to be relevant to the user.

**Comment by JosephTLyons** [comment]
Thanks for reaching out. We want to better support our community, and the best way for us to do that is through our official Discord. The team is active there and happy to help.

We’re closing the “Help and General Q&A” category on GitHub Discussions and directing all support questions to Discord going forward.

https://zed.dev/support
